web/app/js/components/Tap.jsx

```diff
@@ -92,47 +166,264 @@ class Tap extends React.Component {
     this.ws.close(1000);
   }
 
-  handleFormChange = formVal => {
-    let state = {};
-    return e => {
-      state[formVal] = e.target.value;
+  handleFormChange = (name, scopeResource) => {
+    let state = {
+      query: this.state.query,
+      autocomplete: this.state.autocomplete
+    };
+
+    return formVal => {
+      state.query[name] = formVal;
+      if (!_.isNil(scopeResource)) {
+        // scope the available typeahead resources to the selected namespace
+        state.autocomplete[scopeResource] = this.state.resourcesByNs[formVal];
+      }
+
       this.setState(state);
     };
   }
 
+  handleFormEvent = name => {
+    let state = {
+      query: this.state.query
+    };
+
+    return event => {
+      state.query[name] = event.target.value;
+      this.setState(state);
+    };
+  }
+
+  autoCompleteData = name => {
+    return _(this.state.autocomplete[name])
+      .filter(d => d.indexOf(this.state.query[name]) !== -1)
+      .sortBy()
+      .value();
+  }
+
+  loadFromServer() {
+    if (this.state.pendingRequests) {
+      return; // don't make more requests if the ones we sent haven't completed
+    }
+    this.setState({
+      pendingRequests: true
+    });
+
+    let url = "/api/tps-reports?resource_type=all&all_namespaces=true";
+    this.api.setCurrentRequests([this.api.fetchMetrics(url)]);
+    this.serverPromise = Promise.all(this.api.getCurrentPromises())
```

Comment:
TIOLI Can we use the `withREST` component to manage the polling calls that you're adding? I'm not familiar enough with that component to know if would help simplify this component or not.

Reply:
I played around with it a little. It should help simplify some of the boilerplace code like the start/stop polling, etc. However, we do adjust the current state of the various autocompletes a lot so checking props and setting state in the render method (as done in `ResourceList`) won't work here. I think I need to figure out how to incorporate checking withRest differently. I can do that in a follow up.